src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -83,40 +73,26 @@ void CoreWorkerDirectActorTaskSubmitter::ConnectActor(const ActorID &actor_id,
     rpc_clients_[actor_id] =
         std::shared_ptr<rpc::CoreWorkerClientInterface>(client_factory_(address));
   }
-  if (pending_requests_.count(actor_id) > 0) {
-    SendPendingTasks(actor_id);
-  }
+  RAY_CHECK(pending_requests_.find(actor_id) == pending_requests_.end());
+  SendPendingTasks(actor_id);
 }
 
-void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(const ActorID &actor_id,
-                                                         bool dead) {
+void CoreWorkerDirectActorTaskSubmitter::DisconnectActor(const ActorID &actor_id) {
   absl::MutexLock lock(&mu_);
-  if (!dead) {
-    // We're reconstructing the actor, so erase the client for now. The new client
-    // will be inserted once actor reconstruction completes. We don't erase the
-    // client when the actor is DEAD, so that all further tasks will be failed.
-    rpc_clients_.erase(actor_id);
-    worker_ids_.erase(actor_id);
-  } else {
-    RAY_LOG(INFO) << "Failing pending tasks for actor " << actor_id;
-    // If there are pending requests, treat the pending tasks as failed.
-    auto pending_it = pending_requests_.find(actor_id);
-    if (pending_it != pending_requests_.end()) {
-      auto head = pending_it->second.begin();
-      while (head != pending_it->second.end()) {
-        auto request = std::move(head->second);
-        head = pending_it->second.erase(head);
-        auto task_id = TaskID::FromBinary(request->task_spec().task_id());
-        auto status = Status::IOError("cancelling all pending tasks of dead actor");
-        task_finisher_->PendingTaskFailed(task_id, rpc::ErrorType::ACTOR_DIED, &status);
-      }
-      pending_requests_.erase(pending_it);
-    }
-    // No need to clean up tasks that have been sent and are waiting for
-    // replies. They will be treated as failed once the connection dies.
-    // We retain the sequencing information so that we can properly fail
-    // any tasks submitted after the actor death.
-  }
+  DisconnectActorUnprotected(actor_id);
+  // No need to clean up tasks that have been sent and are waiting for
+  // replies. They will be treated as failed once the connection dies.
+  // We retain the sequencing information so that we can properly fail
+  // any tasks submitted after the actor death.
+}
+
+void CoreWorkerDirectActorTaskSubmitter::DisconnectActorUnprotected(const ActorID &actor_id) {
+  rpc_clients_.erase(actor_id);
+  worker_ids_.erase(actor_id);
+  pending_force_kills_.erase(actor_id);
+  pending_requests_.erase(actor_id);
+  next_send_position_.erase(actor_id);
+  next_send_position_to_assign_.erase(actor_id);
```

Comment:
It might be a good idea to have a single map from `ActorID` to a struct holding all of those 🤔